### PR TITLE
Datastore replication engine

### PR DESCRIPTION
**Pull request by Eunovo**
### Description

Prototype Replication Engine which uses Offix-Scheduler to queue mutations. Currently, only "ADD" events can be replicated. "UPDATE" and "DELETE" events will probably require the generation of filter inputs to be sent with inputs to server. Also, proper error handling is still pending.

See #49 for discussions on this work.

##### Checklist

- [x] `npm test` passes
- [x] `npm run build` works
- [x] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by Eunovo** [comment]
@kingsleyzissou  We have bundle size causing tests to fail.

File: packages/offix-datastore/src/replication/ReplicationEngine.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    private storage: Storage;
+
+    constructor(
+        api: IReplicator,
+        storage: Storage
+    ) {
+        this.api = api;
+        this.storage = storage;
+        this.scheduler = new Promise((resolve, reject) => {
+            const scheduler = new OffixScheduler<StoreChangeEvent>({
```
Comment (wtrocki) [inline packages/offix-datastore/src/replication/ReplicationEngine.ts:21]: This looks really ugly. We need scheduler to change. There is no need for init for scheduler in new API etc.

Comment (wtrocki) [reply packages/offix-datastore/src/replication/ReplicationEngine.ts:21]: Again this is more on core team shoulders to improve scheduler. Gianluca works on this and we can see API changed.

**Comment by wtrocki** [comment]
Let's disable size checks for moment on this repo

**Comment by kingsleyzissou** [comment]
@Eunovo @wtrocki we can disable the size-limit stuff, but it's actually quite easy to update now that I've added the config. Eunovo, you can go into the `.size-limit.js` file and just update the config file and set a buffer.

File: packages/offix-datastore/src/replication/ReplicationEngine.ts (hunk 1 above)
Comment (kingsleyzissou) [reply packages/offix-datastore/src/replication/ReplicationEngine.ts:21]: Yeah agreed, this is something that we've needed to improve for a while now.

File: packages/offix-datastore/src/replication/GraphQLCrudQueryBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            const fieldsBuilder: string[] = Object.keys(fields).map((key) => {
+                const graphQLKey = fields[key].key;
+                return graphQLKey;
+            });
+            const graphQLFields = fieldsBuilder.join("\n");
+            const modelName = model.getName();
+
+            const mutations = {
+                create: gql`
+                mutation create${modelName}($input: Create${modelName}Input!) {
```
Comment (kingsleyzissou) [inline packages/offix-datastore/src/replication/GraphQLCrudQueryBuilder.ts:25]: @Eunovo @wtrocki we should probably have some documentation to explain that this is aligned to the graphqlcrud spec.

Comment (wtrocki) [reply packages/offix-datastore/src/replication/GraphQLCrudQueryBuilder.ts:25]: I would use this as a prototype for the moment. We winning and getting lots of progress. Once we get prototype working end to end with all we want to have, we will need to smooth it up and document it

Comment (Eunovo) [reply packages/offix-datastore/src/replication/GraphQLCrudQueryBuilder.ts:25]: Sure, for now, I only add a jsDoc comment on the class to indicate it's purpose

**Comment by Eunovo** [comment]
There's a lot of work to be done here, but we can merge this into next. I'm creating new issues for other features

**Comment by wtrocki** [comment]
Yes. I agree that this is still not prod ready but we should kinda work on individual TODO's together. Amazing work

**Pull request merged**